feat(packer) Adding support for packer templates provided inline

avram opened the pull request:
Adds support for the BakeRequest received by Rosco to have the Packer template inline, as raised in Slack https://spinnakerteam.slack.com/archives/C0DPVDMQE/p1525635535000023

The motivation is that we have teams wanting to be able to customize the Packer usage beyond those components that can be controlled through variables, with details like inline scripts, customized `pause_before` or other details. While we've looked at having a secondary process synchronize custom Packer templates to the Rosco instance, the proposal here is to let pipelines include the template verbatim (potentially in the future sourced from an artifact?) in the bake stage.

I have some concern that this opens up the ability for users with access to creating and running bake stages to run arbitrary provisioners, but I don't see any specific threat there -- anything being run based on the template will run on the instance being provisioned with Packer, not on the Rosco host.

If this looks like a fruitful direction, I'll follow up with changes to Deck and Orca (?) accordingly.

avram commented:
I now see that there is a very similar and more mature concept emerging in #248. I'll take a look at that and see what, if anything, would still need to be done to solve my custom manifest problem.

avram commented:
On further reading, that PR is interesting but doesn't really conflict with this-- it is just about reading the packer output differently.

avram commented on rosco-core/src/main/groovy/com/netflix/spinnaker/rosco/providers/CloudProviderBakeHandler.groovy:
```diff
@@ -212,14 +218,26 @@ abstract class CloudProviderBakeHandler {
+        FileCopyUtils.copy(bakeRequest.template.getBytes(StandardCharsets.UTF_8), tempFile.toFile())
+
+        finalTemplateFileName = tempFile.fileName
+        finalTemplateFilePath = tempFile
```
Since we are pretty far away from the exit of the packer command in this bit of code, I added a line to have the file deleted on VM exit. I think that that should be enough, but let me know if not because it would take some larger reworking to delete immediately on completion of the bake.

ezimanyi replied:
One option to consider is similar to what I did in https://github.com/spinnaker/rosco/pull/253/files, where I delete the temporary file in `completeBake` which is a callback upon completion of the packer command.

avram commented on rosco-core/src/main/groovy/com/netflix/spinnaker/rosco/api/BakeRequest.groovy:
```diff
@@ -75,6 +75,8 @@ class BakeRequest {
   Map extended_attributes
   @ApiModelProperty("The name of a json file containing key/value pairs to add to the packer command (must be in the same location as the template file)")
   String var_file_name
+  @ApiModelProperty("Packer template to use, included inline")
```
Done in ede4938

avram commented:
I've addressed all the feedback so far.

Please wait to merge until I've had a chance to make changes in Deck and elsewhere as needed to make this fully operational.

ezimanyi commented:
Reviewed the latest set of changes---this looks great to me!  I'll wait to hear from you to merge.  (Though looking at the code, it looks like it should be safe to merge this without the changes to the other microservices, at it will just be a no-op if nobody passes in the new `template` parameter.)

duftler commented:
I really like the idea of easing the requirement to pre-position everything in the environment running the packer executable.

Would it be possible to build what you're doing here on top of the recent artifacts work? Might be a nice solution to be able to pass along multiple artifacts, or gzips, or whatever.

That approach might also allow you to reuse lots of existing bits in orca and deck.

avram commented:
> Would it be possible to build what you're doing here on top of the recent artifacts work? Might be a nice solution to be able to pass along multiple artifacts, or gzips, or whatever.

Other PRs are already positioning us to have the packages passed to the default bake script come from artifacts-- not going to touch that right now.

I do think that making it so that the inline template is actually sourced from an artifact will make my life easier to wire into orca/deck, so I'll give that a try.

robzienert commented:
Ping - what's the status of this?

avram commented:
I no longer needed this for my immediate teams' work, and I haven't made progress on it since last spring. I think that it is the right way for Rosco to move and would do a lot to make Rosco + Packer a good experience for Spinnaker operators, but I will probably not be working on this any more in the near future.

robzienert commented:
@avram In that case, I'm going to close this. Of course, feel free to re-open when it's back on your plate.
